### PR TITLE
Issue 163: delete files recursively

### LINKED ISSUE #163: Implement support for deleting directories recursively
# Task Description

We need to implement support for deleting directories recursively in the `org.carlspring.cloud.storage.s3fs.S3FileSystemProvider` class, as this currently only works for regular files, or empty directories and does not recurse into directories, nor remove any files that exist in them.

# Tasks

The following tasks will need to be carried out:
* [ ] Study the code of the `org.carlspring.cloud.storage.s3fs.S3FileSystemProvider` and propose the most efficient way to do this.
* [ ] Implement the necessary changes.
* [ ] Implement test cases.

# Useful links

* [Blog - Multi-Object Delete Announcement](https://aws.amazon.com/about-aws/whats-new/2011/12/07/amazon-s3-announces-multi-object-delete/#:~:text=We%20are%20excited%20to%20announce,your%20application%20and%20Amazon%20S3.)
* [Dev Guide - Multi-Object Delete Docs and Examples](https://docs.aws.amazon.com/AmazonS3/latest/dev/DeletingMultipleObjects.html)
* [Dev Guide - Deleting objects](https://docs.aws.amazon.com/AmazonS3/latest/dev/DeletingObjects.html)
* [Dev Guide - Listing Keys Hierarchically Using a Prefix and Delimiter](https://docs.aws.amazon.com/AmazonS3/latest/dev/ListingKeysHierarchy.html)
* [API Delete Objects](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html)

# Help

* Our [chat channel](https://chat.carlspring.org/channel/s3fs-nio-community)
* Points of contact:
  * @carlspring
  * @ptirador
  * @steve-todorov
  * @sbespalov

### PR DESCRIPTION
**Pull request by dmndZhWng**
# Pull Request Description

This pull request closes #163

# Acceptance Test

* [x] Building the code with `mvn clean install -Pintegration-tests` still works.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [x] Yes, please see [in FileSystemProvider class, the delete(Path) method will not throw DirectoryNotEmptyException anymore]
  * [ ] No

### CODE REVIEW

**Comment by carlspring** [comment]
Hi @dmndZhWng ,

Thank you for you contribution!

Could you please make the test a bit more extensive and include the following:
* Create some files in the directory
* Create a few nested directories with files with them
* Check for the existence of the above files after performing the delete operation

Also, would you mind signing the [ICLA](https://5fdcaf0707fed50007c16708--s3fs-nio.netlify.app/assets/resources/pdfs/ICLA.pdf), as described in the [Contributing](https://s3fs-nio.carlspring.org/contributing/legal/) page?

Also, please, feel free to join our [chat channel](https://chat.carlspring.org/), if you'd like to learn more about the project and/or like to find out what else you could help with.

Kind regards,

Martin

**Comment by dmndZhWng** [comment]
Hi, I ve submit a new PR with some code formatting + fix + Unit Test.
Plus :
1/ I add a new Mock method for submit deleteObjectsRequest (multi request at same time)
2/ I removed the Unit Test for deleteDirectoryWithEntries (because we do not need it anymore since the delete method now handles the case)
However, I d like to know if is there a way to have a technical credential for the Integration test ?

**Comment by carlspring** [comment]
> Hi, I ve submit a new PR with some code formatting + fix + Unit Test.

Thanks for your fixes! We'll review them and get back to you!

> Plus :
> 1/ I add a new Mock method for submit deleteObjectsRequest (multi request at same time)
> 2/ I removed the Unit Test for deleteDirectoryWithEntries (because we do not need it anymore since the delete method now handles the case)

Great! Thank you very much!

> However, I d like to know if is there a way to have a technical credential for the Integration test ?

You would have to create a free tier account and test this.

Have you had a chance to sign the ICLA as I mentioned earlier?

Kind regards,

Martin

**Review by steve-todorov — CHANGES_REQUESTED** [review]
@dmndZhWng thanks for your PR! :) 

I did some digging today and updated the original issue https://github.com/carlspring/s3fs-nio/issues/163. 

The AWS API has limitations - `1000` objects per `listing` and `deletion`.
With the proposed implementation you will be deleting only the first 1000 objects and ignoring the rest. What you should do is use a loop and continue deleting until there are no objects in the `path` or an error occurs (@carlspring should we fail or just print a warning?).

**Comment by carlspring** [comment]
> With the proposed implementation you will be deleting only the first 1000 objects and ignoring the rest. What you should do is use a loop and continue deleting until there are no objects in the `path` or an error occurs (@carlspring should we fail or just print a warning?).

There is no point in warning, or throwing an error here. It should just loop until there are no more.

**Comment by steve-todorov** [comment]
We should at least print warnings with potential exceptions or error messages returned by the API (if any)

**Comment by dmndZhWng** [comment]
I ve made some small changes concerning my previous dev : 

1/ I successfully made an Integration test too :).

2/ I replaced the Files.walk by an recursion call because the documentation says that we cannot rely on the order returned (but the order of path to delete has a real importance in my case) 

3/ I added the delete by batch logic

4/ Concerning the error/warning handling, I notice that the only error thrown is a NoSuchFileException, it happened only when we provide an inexisting path.

**Comment by carlspring** [comment]
Thanks, @dmndZhWng !

@ptirador , would you like to review and test this? :)

**Review by ptirador — CHANGES_REQUESTED** [review]
HI @dmndZhWng! Thanks for your PR! Could you please fix the changes that I requested?
It woukd also be great to create an integration test in `FilesITTest`class, something similar to `deleteDir` method but with filled directories instead of empty.
Thanks.

File: src/test/java/org/carlspring/cloud/storage/s3fs/util/S3ClientMock.java
Comment (ptirador) [reply src/test/java/org/carlspring/cloud/storage/s3fs/util/S3ClientMock.java:20]: Reminder about the imports :)

**Review by ptirador — CHANGES_REQUESTED** [review]

**Comment by ptirador** [comment]
Ping @dmndZhWng! Did you have a chance to look at the changes that I requested?
Thanks!

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/DeleteTest.java
Comment (dmndZhWng) [reply src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/DeleteTest.java:75]: Sorry for answer you so late, i ve already changed this one :).

**Comment by dmndZhWng** [comment]
@ptirador  sorry for my late answer, i just seen your last request concerning the import file.*, and i ve just made the change. please have a look and do not hesitate if any missing. thanks ;).

**Review by ptirador — APPROVED** [review]
LGTM @dmndZhWng! Thank you so much.
cc @carlspring @steve-todorov

**Review by steve-todorov — APPROVED** [review]
LGTM! Thank you so much for your work on this and @ptirador for reviewing! :) 

We will merge this after #184 since currently the tests are not being executed.

**Comment by dmndZhWng** [comment]
@steve-todorov ,@ptirador I have a question concerning the point 4. of what you mentioned in the Issue. 
Because since we cannot pass a boolean `forceContinue` as argument in the delete method (because it inherits from FileSystemProvider).

Maybe we could :
- throw the SdkException at S3FileSystemProvider::delete level (hence, let the user to decide how to handle that case) ?
- or wrap the SDKException and throw it as an IOException in order to keep only 1 base exception for the delete method ?

Because in the current version, we might get directly a SDKException if something happens when delete ...

Do not hesitate if you have any idea / preference.

**Comment by steve-todorov** [comment]
Oh sorry, I missed that question! 

If the thrown exception is `NoSuchFileException` then you can just `catch` it and add a debug log with something like `Deleting [/path/to/file] was skipped because the path was not found.`. For other cases I think we should throw an `IOException` as soon as the error happens. 

@ptirador / @carlspring thoughts?

**Comment by dmndZhWng** [comment]
I think maybe this case has to be handled with the same way too (i.e. log and skipped) ? 
`try (final DirectoryStream<Path> stream = Files.newDirectoryStream(s3Path))` because `Files.newDirectoryStream` might throw a `SecurityException` in the case of the default provider, and a security manager is installed, the `SecurityManager#checkRead(String) checkRead` method is invoked to check read access to the directory.
@steve-todorov / @ptirador / @carlspring thoughts?

**Comment by ptirador** [comment]
> Oh sorry, I missed that question!
> 
> If the thrown exception is `NoSuchFileException` then you can just `catch` it and add a debug log with something like `Deleting [/path/to/file] was skipped because the path was not found.`. For other cases I think we should throw an `IOException` as soon as the error happens.
> 
> @ptirador / @carlspring thoughts?

I agree with this solution. Thanks :)

File: src/main/java/org/carlspring/cloud/storage/s3fs/S3FileSystemProvider.java
Comment (ptirador) [inline src/main/java/org/carlspring/cloud/storage/s3fs/S3FileSystemProvider.java:813]: Format specifiers should be used instead of string concatenation.

```suggestion
            LOGGER.warn("Deleting {} was skipped because the path was not found.", s3Path);
```

File: src/main/java/org/carlspring/cloud/storage/s3fs/S3FileSystemProvider.java
Comment (ptirador) [inline src/main/java/org/carlspring/cloud/storage/s3fs/S3FileSystemProvider.java:830]: Format specifiers should be used instead of string concatenation.

```suggestion
                    LOGGER.warn("Deleting {} was skipped because the path could not be read-accessed.", s3Path);
```

**Review by ptirador — CHANGES_REQUESTED** [review]
Hi @dmndZhWng , could you take a look at these small changes?
Thanks!

**Comment by dmndZhWng** [comment]
hello done :) @ptirador . do not hesitate if any other questions

**Review by ptirador — APPROVED** [review]
LGTM @dmndZhWng! Thanks.
cc @carlspring

**Comment by carlspring** [comment]
Thanks for reviewing, @ptirador !

This will be merged once @steve-todorov had merged #184 and #186 and tested this pull request.

**Comment by steve-todorov** [comment]
I have `rebased`, `squashed` and `tested`. The S3 failure in GH Actions because `secrets` are not passed down forks (known GH Actions issue). I have manually tested this code with S3 and it's working as expected! :) 

Thanks to everybody for your active work on this! I'm proceeding with merging into the `master` :)

**Comment by carlspring** [comment]
@dmndZhWng ,

Thank you very much for your contribution! Would you like to have a look at #160 as well?

@ptirador , @steve-todorov : Thanks for reviewing and merging this! :)

**Comment by dmndZhWng** [comment]
@carlspring yes sure, there are some common logic i think

**Comment by carlspring** [comment]
Excellent! Thank you!

**Pull request merged**
